static/app/components/events/eventTags/eventTagsTree.tsx

```diff
@@ -0,0 +1,202 @@
+import styled from '@emotion/styled';
+
+import EventTagsContent from 'sentry/components/events/eventTags/eventTagContent';
+import {space} from 'sentry/styles/space';
+import type {EventTag} from 'sentry/types';
+import {generateQueryWithTag} from 'sentry/utils';
+import {useLocation} from 'sentry/utils/useLocation';
+import useOrganization from 'sentry/utils/useOrganization';
+
+const MAX_TREE_DEPTH = 4;
+interface TagTree {
+  [key: string]: TagTreeContent;
+}
+interface TagTreeContent {
+  subtree: TagTree;
+  value: string;
+  // These will be omitted on pseudo tags (see addToTagTree)
+  meta?: Record<any, any>;
+  originalTag?: EventTag;
+}
+
+function addToTagTree(
+  tree: TagTree,
+  tag: EventTag,
+  meta: Record<any, any>,
+  originalTag: EventTag
+): TagTree {
+  const branchMatches = tag.key.match(/\./g) ?? [];
+
+  const hasInvalidBranchCount =
+    branchMatches.length <= 0 || branchMatches.length > MAX_TREE_DEPTH;
+  const hasInvalidBranchSequence = (tag.key.match(/\.{2,}/g) ?? []).length > 0;
```

Comment:
probably want to use test
```suggestion
  const hasInvalidBranchSequence = /\.{2,}/g.test(tag.key)
```

Reply:
TIL 🙏